engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/NadelServiceTypeFilterTransform.kt

```diff
@@ -0,0 +1,144 @@
+        if (typesOwnedByService.size == overallField.objectTypeNames.size) {
+            return null
+        }
+
+        return State(
+            aliasHelper = NadelAliasHelper.forField(
+                tag = "type_filter",
+                field = overallField,
+            ),
+            typeNamesOwnedByService = typeNamesOwnedByService,
+            fieldObjectTypesOwnedByService = typesOwnedByService,
+            overallField = overallField,
+        )
+    }
+
+    override suspend fun transformField(
+        executionContext: NadelExecutionContext,
+        transformer: NadelQueryTransformer.Continuation,
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        service: Service,
+        field: ExecutableNormalizedField,
+        state: State,
+    ): NadelTransformFieldResult {
+        // Nothing to query if there are no fields, we need to add selection
+        if (state.fieldObjectTypesOwnedByService.isEmpty()) {
+            val objectTypeNames = state.overallField.parent.getFieldDefinitions(executionBlueprint.schema)
+                .asSequence()
+                .flatMap { fieldDef ->
+                    resolveObjectTypes(
+                        executionBlueprint.schema,
+                        type = fieldDef.type,
+                        onNotObjectType = { type ->
+                            error("Unable to resolve to object type: $type")
+                        }
+                    )
+                }
+                .map {
+                    it.name
+                }
+                .filter {
+                    it in state.typeNamesOwnedByService
+                }
+                .toSet()
+                .toList()
+
+            return NadelTransformFieldResult(
+                newField = null,
+                artificialFields = listOf(
+                    newNormalizedField()
+                        .objectTypeNames(objectTypeNames)
+                        .alias(state.aliasHelper.typeNameResultKey)
+                        .fieldName(Introspection.TypeNameMetaFieldDef.name)
+                        .build(),
+                ),
+            )
+        }
+
+        return NadelTransformFieldResult(
+            newField = field
+                .toBuilder()
+                .clearObjectTypesNames()
+                .objectTypeNames(state.fieldObjectTypesOwnedByService)
+                .children(transformer.transform(field.children))
+                .build(),
+        )
+    }
+
+    override suspend fun getResultInstructions(
+        executionContext: NadelExecutionContext,
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        service: Service,
+        overallField: ExecutableNormalizedField,
+        underlyingParentField: ExecutableNormalizedField?,
+        result: ServiceExecutionResult,
+        state: State,
+    ): List<NadelResultInstruction> {
+        return emptyList()
+    }
+
+    private fun getOwnedTypeNames(
```

Comment:
will this function return **all** types owned by the service OR **all types declared by the field** that are owned by the service?

Reply:
I'll rename the function. All types owned by a service.

Reply:
Right, so in that case, should we be checking that all items in `overallField.objectTypeNames` are present in `typesOwnedByService` instead of just comparing their sizes?

```
        if (typesOwnedByService.size == overallField.objectTypeNames.size) {
            return null
        }
```

Like, if we have this:

```
typesOwnedByService = [Issue, TodoIssue, PullRequest]
overallField.objectTypeNames = [Issue, TodoIssue]
```

we still want to return `null` here, right?

Reply:
Oh gosh that code is a mess. Sorry I didn't notice that after refactoring. There's actually two variables there `typeNamesOwnedByService` and `typesOwnedByService`. I've refactored it to the following now

```kotlin
        val typeNamesOwnedByService = getTypeNamesServiceOwns(executionBlueprint, service)
        val fieldObjectTypeNamesOwnedByService = overallField.objectTypeNames
            .filter { it in typeNamesOwnedByService }

        // All types are owned by service
        if (fieldObjectTypeNamesOwnedByService.size == overallField.objectTypeNames.size) {
            return null
        }
```